ADR/0039-workspace-deprecation.md

```diff
@@ -0,0 +1,98 @@
+
+* We had our [first upstream Konflux release](https://github.com/konflux-ci/konflux-ci) which doesn't use [kubesaw] or relying
+on the Workspace abstraction, and is deployable on vanilla Kubernetes.
+
+* We made a decision (now long ago) to drop [KCP] from our architecture, which initially introduced the workspace concept. Both transparent multi-cluster as a feature and workspaces as a concept came from KCP.
+
+* We made the decision (at Red Hat) to stop trying to offer a managed service interface for Konflux. The user signup, waitlist, and workspace provisioning interfaces made a lot of sense there.
+
+## Decision
+
+1. We will remove the Workspace abstraction from Konflux and use standard Kubernetes namespaces
+directly for providing a place for users to run their pipelines and store their configurations
+and secrets.
+
+2. We will stop using the term Workspace, and start to use the term Namespace.
+
+3. The Konflux UI will expose a wizard for creating a new namespace. This
+wizard will be available to any user that has permissions to create namespaces.
+
+4. Konflux will provide an optional automation for creating a namespace for a user
+when he/she logins for the first time.
+
+5. Konflux won't be opinionated about the mechanism for initializing/maintaining namespaces with
+supporting resources such as ResourceQuota and LimitRange. Konflux will differ
+this responsibility to other tools which are specialized in this task such as 
+(but not limited to) [Kyverno](https://kyverno.io/policies/best-practices/add-ns-quota/add-ns-quota/) and [ArgoCD](https://github.com/konflux-ci/namespace-generator).
+
+6. Konflux will provide a thin [backend service](https://github.com/konflux-ci/workspace-manager) for listing the namespaces where the user has at least view access
+to the Konflux CRDs. This list will be used by the namespace switcher in the UI.
+This is required since the Kubernetes API doesn't let the user to list a subset
+of namespace. The user gets permissions to list all namespace or none.
+
+7. Konflux will provide ClusterRoles that will grant permissions to the Konflux
+and Tekton CRDs. Those will be (aggregated)[https://github.com/konflux-ci/konflux-ci/issues/440] to the built-in Kubernetes roles (`view`, `edit`, `admin`).
+
+8. Public viewer access will be provided by assigning the `view` role (see above) to the
+`system:authenticated` built-in group that contains all the authenticated users.
+
+9. Same as Kubernetes, Konflux won't have a resource for representing a user. Instead,
+it will use external Identity providers.
+
+9. Konflux won't provide a way for creating a ephemeral namespaces. It will differ this
+task to another tool.
+
+10. The `join the waitlist` button will be removed from the Konflux UI.
+
+11. Konflux won't provide a transparent multi-cluster deployment.
+If required, existing open source projects that handle multi-cluster deployments
+should be explored and Konflux should integrate with them.
```

Comment:
> What benefits are we losing if we no longer have the ability to support multi-cluster?

Users will need to open multiple tabs in the browser if working with multiple namespace on different clusters, since each UI instance will show only the namespaces on the cluster it was deployed to.

> If we need to support a multi-cluster use case in the future, we will need to re-add an abstraction like this. How might that compare to a Workspace abstraction? How big will the change be for this (will it largely just be a UI change)?

It won't be a big change. It would require adding the workspace name to each request and change the namespace lister to know how to parse Workspace objects.

If it's a concern, we can keep the UI as is, and remove the workspace part from the path of the requests and also make the service that lists namespaces return a response that contains a list of workspace objects (the list is generated on the fly from the list of namespace the user has access to). This is the way konflux-ci/konflux-ci works today (without the workspace abstraction https://github.com/konflux-workspaces/workspaces)

Reply:
but you don't have to use it if you don't want to. The proxy that provides transparent multi-cluster (and all features around it) was implemented for Konflux mainly. 

> If we need to support a multi-cluster use case in the future, we will need to re-add an abstraction like this. How might that compare to a Workspace abstraction? How big will the change be for this (will it largely just be a UI change)? 

You can use KubeSaw in single-cluster mode and extend it later if you want - there are no costs, and we already did that in DevSandbox (it was originally single-cluster environment)

> Users will need to open multiple tabs in the browser if working with multiple namespace on different clusters, since each UI instance will show only the namespaces on the cluster it was deployed to.

This is already possible with KubeSaw - DevSandbox works like that,  we don't use proxy there at all.

Reply:
If the are no objections I'll close this thread.